torch/distributed/elastic/timer/file_based_local_timer.py

```diff
@@ -68,69 +64,67 @@ class FileTimerClient(TimerClient):
-        try:
-            fd = os.open(self._file_path, os.O_WRONLY | os.O_NONBLOCK)
-            return os.fdopen(fd, "wt")
-        except Exception:
-            return None
-
-    def _send_request(self, request: FileTimerRequest) -> None:
-        # The server may have crashed or may haven't started yet.
-        # In such case, calling open() in blocking model blocks the client.
-        # To avoid such issue, open it in non-blocking mode, and an OSError will
-        # be raised if the server is not there.
-        file = self._open_non_blocking()
-        if file is None:
-            raise BrokenPipeError("Could not send the FileTimerRequest because FileTimerServer is not available.")
-        with file:
+        log.info(
+            "Starting %s..."
+            " file_path=%s,"
+            " timeout=%s,"
+            " signal=%s",
+            type(self).__name__,
+            self._file_path,
+            self._timeout_sec,
+            self.signal,
+        )
+        with open(self._file_path, "w") as file:
+            # FileTimerRequest at init is used to write client level
+            # metadata once to watchdog file, later acquire will not
+            # write duplicate data
+            request = FileTimerRequest(
+                worker_pid=os.getpid(),
+                scope_id=scope_id,
+                expiration_time=timeout_sec,
+                signal=signal,
+            )
             json_request = request.to_json()
-            # Write request with no greater than select.PIPE_BUF is guarantee to be atomic.
-            if len(json_request) > select.PIPE_BUF:
-                raise RuntimeError(
-                    f"FileTimerRequest larger than {select.PIPE_BUF} bytes "
-                    f"is not supported: {json_request}"
-                )
             file.write(json_request + "\n")
 
     def acquire(self, scope_id: str, expiration_time: float) -> None:
-        self._send_request(
-            request=FileTimerRequest(
-                worker_pid=os.getpid(),
-                scope_id=scope_id,
-                expiration_time=expiration_time,
-                signal=self.signal
-            ),
-        )
+        if not os.path.isfile(self._file_path):
+            raise FileNotFoundError("Could not send the FileTimerRequest because FileTimerServer is not available.")
+        # updating watchdog file modified time to current time as heartbeat
+        # avoid writing duplicate FileTimerRequest, expiration time is
+        # calcuated using 'modified time + timeout'
+        os.utime(self._file_path)
 
     def release(self, scope_id: str) -> None:
-        self._send_request(
-            request=FileTimerRequest(
-                worker_pid=os.getpid(),
-                scope_id=scope_id,
-                expiration_time=-1,
-                signal=0
-            ),
-        )
+        # release is no-op as timer are alive until end of server shutdown
```

Comment:
what about for short lived watchdog blocks?
This is very much intended for arbitrary numbers of small watchdog sections and part of the public api [https://pytorch.org/docs/stable/elastic/timer.html](https://l.facebook.com/l.php?u=https%3A%2F%2Fpytorch.org%2Fdocs%2Fstable%2Felastic%2Ftimer.html&h=AT371X5StE7-oG1SH1gCU_dvcdaC-ds14kgSrBY9tccwaxvH6F_0gBJAwqynDImZi8xRhneoaACfq5GD48Y46omJulpde4yt7CWX2Cfb9oQV3GgenoXbFOVVBrQ)

Reply:
with new implementation we are not offering short lived watchdog blocks with FileBasedTimer, goal is to have simplified watchdog for the whole duration of client worker

should we keep this as is for BC, and create altogether new classes/methods for new approach? WDYT?